sys/exec.c

```diff
@@ -107,12 +107,35 @@ int do_exec(const exec_args_t *args) {
   vm_map_t *old_vmap = vm_map_activate(vmap);
 
   /* Iterate over prog headers */
-  log("ELF has %d program headers", eh->e_phnum);
-
-  const uint8_t *phs_base = elf_image + eh->e_phoff;
+  log("ELF has %d program headers", eh.e_phnum);
+
+  assert(eh.e_phoff < 64);
+  assert(eh.e_phentsize < 128);
+  /* We allocate this buffer on stack. It's confirmed to be small. A statically
+     allocated buffer won't do, because it would need restrictive
+     synchronization. Using a malloc pool would be probably an overkill for this
+     case. */
+  size_t phs_size = eh.e_phoff * eh.e_phentsize;
+  char phs[phs_size];
+
+  /* Read program headers. */
+  uio.uio_op = UIO_READ;
```

Comment:
This piece of code looks familiar and apparently occurs in several places in the kernel. Could we wrap it with `uio_make(&uio, UIO_READ, ...)` function?